torch/csrc/api/src/optim/lbfgs.cpp

```diff
@@ -30,139 +68,191 @@ Tensor LBFGS::gather_flat_grad() {
       views.push_back(parameter.grad().view(-1));
     }
   }
-  return torch::cat(views);
+  return torch::cat(views, 0);
 }
 
-void LBFGS::add_grad(const torch::Tensor& step_size, const Tensor& update) {
+int64_t LBFGS::_numel() {
+  if(_numel_cache == c10::nullopt) {
+    auto res = 0;
+    for (auto& parameter : _params) {
+      int64_t numel = parameter.numel();
+      res+=numel;
+    }
+    *_numel_cache = res;
+  }
+  return *_numel_cache;
+}
+
+void LBFGS::_add_grad(const double step_size, const Tensor& update) {
   NoGradGuard guard;
   int64_t offset = 0;
-  for (auto& parameter : parameters_) {
+  for (auto& parameter : _params) {
     int64_t numel = parameter.numel();
     parameter.add_(
-        update.slice(0, offset, offset + numel, 1).view_as(parameter),
-        step_size.item<float>());
+      update.slice(0, offset, offset + numel, 1).view_as(parameter), step_size);
     offset += numel;
   }
+  TORCH_INTERNAL_ASSERT(offset == _numel())
 }
 
-torch::Tensor LBFGS::step(LossClosure closure) {
-  torch::Tensor orig_loss = closure();
-  torch::Tensor loss = orig_loss.clone(at::MemoryFormat::Contiguous);
-  int64_t current_evals = 1;
-  func_evals += 1;
+Tensor LBFGS::step(LossClosure closure) {
+  NoGradGuard guard;
+  TORCH_CHECK(closure != nullptr, "LBFGS requires a closure function");
+  TORCH_INTERNAL_ASSERT(param_groups_.size() == 1);
+  auto& group = param_groups_[0];
```

Comment:
I believe we still need
```python
# Make sure the closure is always called with grad enabled
closure = torch.enable_grad()(closure)
```
We can do that by wrapping `torch::AutoGradMode enable_grad(true); closure();` in another `LossClosure` function, and assign it to `closure`.

Reply:
This was done just to make it easier to write in a python side.
You can also just enable grad at the two places where the closure is used.